core/templates/pages/moderator-page/moderator-page.component.ts

```diff
@@ -16,124 +16,145 @@
-        ctrl.allCommits = [];
-        ctrl.allFeedbackMessages = [];
-        // Map of exploration ids to objects containing a single key: title.
-        ctrl.explorationData = {};
-
-        ctrl.displayedFeaturedActivityReferences = [];
-        ctrl.lastSavedFeaturedActivityReferences = [];
-        ctrl.FEATURED_ACTIVITY_REFERENCES_SCHEMA = {
-          type: 'list',
-          items: {
-            type: 'dict',
-            properties: [{
-              name: 'type',
-              schema: {
-                type: 'unicode',
-                choices: [ENTITY_TYPE.EXPLORATION, ENTITY_TYPE.COLLECTION]
-              }
-            }, {
-              name: 'id',
-              schema: {
-                type: 'unicode'
-              }
-            }]
-          }
-        };
-
-        var RECENT_COMMITS_URL = (
-          '/recentcommitshandler/recent_commits' +
-          '?query_type=all_non_private_commits');
-        // TODO(sll): Update this to also support collections.
-        $http.get(RECENT_COMMITS_URL).then(function(response) {
-          // Update the explorationData object with information about newly-
-          // discovered explorations.
-          var data = response.data;
-          var explorationIdsToExplorationData = data.exp_ids_to_exp_data;
-          for (var expId in explorationIdsToExplorationData) {
-            if (!ctrl.explorationData.hasOwnProperty(expId)) {
-              ctrl.explorationData[expId] = (
-                explorationIdsToExplorationData[expId]);
-            }
-          }
-          ctrl.allCommits = data.results;
-          LoaderService.hideLoadingScreen();
-        });
-
-        $http.get('/recent_feedback_messages').then(function(response) {
-          ctrl.allFeedbackMessages = response.data.results.map(
-            d => ThreadMessage.createFromBackendDict(d));
-        });
-
-        $http.get('/moderatorhandler/featured').then(function(response) {
-          ctrl.displayedFeaturedActivityReferences = (
-            response.data.featured_activity_references);
-          ctrl.lastSavedFeaturedActivityReferences = angular.copy(
-            ctrl.displayedFeaturedActivityReferences);
-        });
-      };
-    }
-  ]
-});
+    return path + '/' + reference.id;
+  }
+
+  isSaveFeaturedActivitiesButtonDisabled(): boolean {
+    return _.isEqual(
+      this.displayedFeaturedActivityReferences,
+      this.lastSavedFeaturedActivityReferences);
+  }
+
+  saveFeaturedActivityReferences(): void {
+    this.alertsService.clearWarnings();
+
+    let activityReferencesToSave =
+      [...this.displayedFeaturedActivityReferences];
+
+    this.moderatorPageBackendApiService
+      .saveFeaturedActivityReferences(activityReferencesToSave).then(() => {
+        this.lastSavedFeaturedActivityReferences = activityReferencesToSave;
+        this.alertsService.addSuccessMessage('Featured activities saved.');
+      });
+  }
+
+  getSchema(): Schema {
```

Comment:
If the schema is being served via a function then why we need to attach the constant  FEATURED_ACTIVITY_REFERENCES_SCHEMA to the class, can we directly return the dict?

Reply:
We haven't migrated schema-based-editor yet. So, we use upgradeComponent to use schema-based-editor in our angular components. angularjs directives get the values of passed down variables by calling them like a function. This calling has no side effects for the following cases:
1. ajs directive used in ajs
2. angular component used in ajs

but when we use angularjs directive in the angular component, we cannot directly pass down the value, we have to wrap it around a function.

This will not be needed after schema-based-edtior is migrated. After the migration of schema-based-editor, FEATURED_ACTIVITY_REFERENCES_SCHEMA will directly be passed down to schema-based-editor.

So, it makes more sense to leave the constant as it is. As it will be used after the migration.

Reply:
Actually, we can pass it as an object... but that would mean we would have to change the bindToController in  schema-based-editor from `schema: '&',` to `schema: '<',`. Afraid that this might be used in $comple somewhere, I have asked this not to be changed for now.